src/host-config/proxy.ts

```diff
@@ -0,0 +1,275 @@
+	): string {
+		const lines = Object.entries(block).map(([key, value]) => {
+			if (isAuthField(key)) {
+				// Add double quotes around login and password fields
+				value = `${value.startsWith('"') ? '' : '"'}${value}${value.endsWith('"') ? '' : '"'}`;
+			}
+			return `\t${key} = ${value};`;
+		});
+		return `${label} {\n${lines.join('\n')}\n}\n`;
+	}
+
+	/**
+	 * Given the raw contents of a block redsocks.conf file,
+	 * extract to a key-value object.
+	 */
+	private static parseBlock(
+		rawBlockConf: string,
+		unsupportedKeys: string[],
+	): Record<string, string> {
+		const parsedBlock: Record<string, string> = {};
+
+		// Split by newline and optional semicolon
+		for (const line of rawBlockConf.split(/;?\n/)) {
+			if (!line.trim().length) {
+				continue;
+			}
+			let [key, value] = line.split(/ *?= *?/).map((s) => s.trim());
+			// Don't parse unsupported keys
+			if (key && unsupportedKeys.some((k) => key.match(k))) {
+				continue;
+			}
+			if (key && value) {
+				if (isAuthField(key)) {
+					// Remove double quotes from login and password fields for readability
+					value = value.replace(/"/g, '');
+				}
+				parsedBlock[key] = value;
+			} else {
+				// Skip malformed lines
+				log.warn(
+					`Ignoring malformed redsocks.conf line ${isAuthField(key) ? `"${key}"` : `"${line.trim()}"`} due to missing key, value, or "="`,
+				);
+			}
+		}
+
+		return parsedBlock;
+	}
+}
+
+export async function readProxy(): Promise<HostProxyConfig | undefined> {
+	// Get and parse redsocks.conf
+	let rawConf: string | undefined;
+	try {
+		rawConf = await readFromBoot(redsocksConfPath, 'utf-8');
+	} catch (e: unknown) {
+		if (!isENOENT(e)) {
+			throw e;
+		}
+		return undefined;
+	}
+	const redsocksConf = RedsocksConf.parse(rawConf);
+
+	// Get and parse no_proxy
+	const noProxy = await readNoProxy();
+
+	// Build proxy object
+	const proxy = {
+		...redsocksConf.redsocks,
+		...(noProxy.length && { noProxy }),
+	};
+
+	// Assumes mandatory proxy config fields (type, ip, port) are present,
+	// even if they very well may not be. It is up to the user to ensure
+	// that all the necessary fields are present in the redsocks.conf file.
+	return proxy as HostProxyConfig;
+}
+
+export async function setProxy(
+	conf: RedsocksConfig,
+	noProxy: Nullable<string[]>,
+) {
+	// Ensure proxy directory exists
+	await mkdirp(proxyBasePath);
+
+	// Set no_proxy
+	let noProxyChanged = false;
+	if (noProxy != null) {
+		noProxyChanged = await setNoProxy(noProxy);
+	}
+
+	// If target is empty aside from noProxy and noProxy got patched,
+	// do not change redsocks.conf to remain backwards compatible
+	if (Object.keys(conf).length === 0 && noProxyChanged) {
+		// Restart services using dbus
+		await restartProxyServices();
+		return;
+	}
+
+	// Write to redsocks.conf
+	const toWrite = RedsocksConf.stringify(conf);
+	if (toWrite) {
+		await writeToBoot(redsocksConfPath, toWrite);
+	} else {
+		await unlinkAll(redsocksConfPath);
+	}
```

Comment:
This simplification might work?

```suggestion
	// Write to redsocks.conf
	const toWrite = RedsocksConf.stringify(conf);
	if (toWrite) {
		await writeToBoot(redsocksConfPath, toWrite);
	} else if (!noProxyChanged) {
		await unlinkAll(redsocksConfPath);
	}
```